todo_widget: Disable /todo add task button if duplicate task name or no task name.

Sarah-Addo opened the pull request:
This PR fixes https://github.com/zulip/zulip/issues/20211.

**Testing plan:**
Visual testing by trying to interact with add task button with no task name present and checking that it does get disabled with tooltip. Visual testing with duplicate task name and interacting with button to see that it is disabled with tooltip.

**GIFs or screenshots:** 
GIF of no task name present.
https://user-images.githubusercontent.com/36671046/143097792-34c83b54-6f1b-4858-a63b-16a05644a3cd.mp4

GIF of duplicate task name.
https://user-images.githubusercontent.com/36671046/143097808-ad01c77a-5017-4ed2-82e9-5016eb821d8c.mp4
<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

alya commented:
Hi @Sarah-Addo !

The experience is not quite right for blank tasks. The button should be disabled all the time in that case. In the current implementation, it only becomes disabled when you hover over it.
![Screen Shot 2021-11-17 at 11 26 20 AM](https://user-images.githubusercontent.com/2090066/142269045-fe0309aa-7370-40d2-b172-3b52780e930e.png)

alya commented:
Actually, hm, after I wrote this comment, I realized I'm not so sure!

I think the button greying out when you hover over it looks odd, but it may also be confusing to have it greyed out to begin with.

alya commented:
@timabbott for the tooltips, should we be using tippy here?

timabbott commented:
Yes

-Tim Abbott (mobile)

On Wed, Nov 17, 2021, 11:30 Alya Abbott ***@***.***> wrote:

> @timabbott <https://github.com/timabbott> for the tooltips, should we be
> using tippy here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip/pull/20261#issuecomment-971905414>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAU6NWVYGRU7IJPR5QRHOITUMP7HBANCNFSM5ICT3U3Q>
> .
>

Sarah-Addo commented:
> Actually, hm, after I wrote this comment, I realized I'm not so sure!
> 
> I think the button greying out when you hover over it looks odd, but it may also be confusing to have it greyed out to begin with.

I also had the same thought about it being confusing to have it grayed out to begin with. The button could be disabled upon creation of the todo list with a tooltip that instantly appears? I am currently looking into tippy for the tooltips but if you feel like there are any examples that would be helpful feel free to link them.

alya commented:
We use tippy for a lot of tooltips throughout the app, so you should definitely be able to find some examples.

alya commented:
> I also had the same thought about it being confusing to have it grayed out to begin with. The button could be disabled upon creation of the todo list with a tooltip that instantly appears? 

Hm, that seems like it might be unnecessarily invasive? Maybe the way you implemented is actually the best option. :)

Sarah-Addo commented:
@alya I've changed the tooltips to tippy tooltips and have updated the GIFs linked!

alya commented:
Cool, thanks!

Playing with it a bit more, let's try making the disabled button look like the disabled buttons in our settings menus. For example, this button in the Organization profile tab is disabled if you're not an org owner:

![Screen Shot 2021-11-23 at 8 41 25 AM](https://user-images.githubusercontent.com/2090066/143066554-063343f1-42af-494c-87ac-a5dba1307f38.png)

If you play with it, you'll see that the tooltip is different from an active button, and the text color changes when you hover over it.

Sarah-Addo commented:
I've styled the disabled button accordingly but I was wondering if you would like the disabled cursor symbol to appear since it seems to overlap with the tippy tooltip depending on how the user hovers over it which you can see [here](https://user-images.githubusercontent.com/36671046/143087257-9141d6ec-0025-4485-9766-09dd8f41356e.mp4)

alya commented:
I think the overlap isn't too bad, and it's better to be consistent.

Let me know when it's ready for another review. :)

Sarah-Addo commented:
Ready for another review!

alya commented:
Cool, thanks! The button looks good, but I noticed that the tooltip sometimes gets cut off. Is it somehow behind an element that it should be in front of?
<img width="834" alt="Screen Shot 2021-11-23 at 3 58 43 PM" src="https://user-images.githubusercontent.com/2090066/143147385-6b90d465-9252-4fa7-972f-2745c7fd2179.png">

Sarah-Addo commented:
Yep it looks like when the list was empty the absolute positioning of the element was causing the cut off. I changed the positioning to relative and updated the css to ensure that the button still has the disabled effects when you hover over the wrapper which you can see [here](https://user-images.githubusercontent.com/36671046/143180483-b75fa7ef-b1d7-4611-9da7-ea0ade5dfdbd.mp4)

alya commented:
Great! @timabbott this PR is be ready for a code review.

timabbott commented on static/styles/widgets.css:
```diff
@@ -167,3 +175,20 @@ button {
+    z-index: 1;
+
+    &:hover {
+        cursor: not-allowed;
```
Hmm, why is the `cursor` declaration here?  It seems like we wouldn't not this cursor in the case it isn't disabled.

Sarah-Addo replied:
When testing and using the todo, it was evident that the styling of the enabled button trumped the styling of the div so it functioned [normally](https://user-images.githubusercontent.com/36671046/143785490-87a6dbeb-11c5-4707-a016-e705525b22ea.mp4).

timabbott commented on static/styles/widgets.css:
```diff
@@ -167,3 +175,20 @@ button {
+.add-task-wrapper {
+    display: inline;
+    position: relative;
+    z-index: 1;
```
Why is `z-index` needed here?

Sarah-Addo replied:
When the button becomes disabled the z-index is set to -1. When testing it was evident that the wrapper needed to be strictly "on top" of the button in order for the tippy tooltip to be triggered ([this](https://user-images.githubusercontent.com/36671046/143785928-c6c309f9-8cec-49c0-8b4b-fe8bd8769ad1.mp4) is without z-index). This is also why the additional grayed out styling of the disabled button is determined by the hovering of the wrapper because it's now "on top".

timabbott commented on static/styles/widgets.css:
```diff
@@ -114,6 +114,14 @@ button {
+            filter: saturate(0);
+            background-color: hsl(0, 0%, 93%);
+            color: hsl(0, 3%, 52%);
+        }
```
Do we need night theme CSS for this as well?

Sarah-Addo replied:
Could you expand on this? I essentially copied the css for this disabled button
![image](https://user-images.githubusercontent.com/36671046/143787444-8f012589-aa8d-4864-9795-26306ff60754.png)

timabbott commented on static/js/todo_widget.js:
```diff
@@ -161,30 +161,64 @@ export function activate(opts) {
+                e.stopPropagation();
+                check_valid_task();
+            },
+        );
```
Can you add a comment explaining why this function is needed?

Sarah-Addo replied:
Done!

timabbott commented:
Thanks for working on this @Sarah-Addo!  I posted a few comments; it'd also be great if you can clean up the commit history (possibly by squashing the commits; I haven't tried to read them individually).

Check out our [GitHub guide](https://zulip.readthedocs.io/en/latest/git-guide.html) and [commit guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html) for more details.

Sarah-Addo commented:
I've addressed the comments left and this is ready for another review :)

timabbott commented:
Closing as #20211 was fixed via another PR. Thanks for the effort @Sarah-Addo!
